File: csharp/parsing-log-files/ParsingLogFiles.cs
using System.Text.RegularExpressions;
using System.Text;

public class LogParser
{
    public bool IsValidLine(string text) => Regex.IsMatch(text, @"^\[(TRC|DBG|INF|WRN|ERR|FTL)\].*$");

    public string[] SplitLogLine(string text) => Regex.Split(text, @"<[\^\*\=\-]+>");

    public int CountQuotedPasswords(string lines) => Regex.Matches(lines, @"^.*"".*password.*"".*$", RegexOptions.Multiline | RegexOptions.IgnoreCase).Count;

    public string RemoveEndOfLineText(string line) => string.Join("", new Regex(@"end\-of\-line\d+").Split(line));

    public string[] ListLinesWithPasswords(string[] lines)
    {
        var result = new StringBuilder();
        foreach (var line in lines)
        {
            var passwordsInLine = Regex.Match(line, @".*(password\S+).*$", RegexOptions.IgnoreCase).Groups;
            result = result.AppendLine(formatResultLine(passwordsInLine, line));
        };

        return result.ToString().Trim().Split("\n");
    }

    private string formatResultLine(GroupCollection passwordsInLine, string line)
    {
        return passwordsInLine[1].Success ? $"{passwordsInLine[1].Value}: {line}" : $"--------: {line}";
    }
}

File: go/parsing-log-files/parsing_log_files.go
package parsinglogfiles

import (
	"fmt"
	"regexp"
)

func IsValidLine(text string) bool {
	re := regexp.MustCompile(`^\[(ERR|TRC|DBG|INF|WRN|FTL)\].*`)

	return re.MatchString(text)
}

func SplitLogLine(text string) []string {
	re := regexp.MustCompile(`<[~*=-]*>`)

	return re.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
	re := regexp.MustCompile("(?i)" + `\".*password.*\"`)

	passwordCount := 0
	for i := 0; i < len(lines); i++ {
		if re.FindString(lines[i]) != "" {
			passwordCount++
		}
	}

	return passwordCount
}

func RemoveEndOfLineText(text string) string {
	re := regexp.MustCompile(`end-of-line[\d]+`)

	return re.ReplaceAllString(text, "")
}

func TagWithUserName(lines []string) []string {
	re := regexp.MustCompile(`User[ ]+([\w]*)`)
	results := make([]string, len(lines))
	for i := 0; i < len(lines); i++ {
		matches := re.FindStringSubmatch(lines[i])
		userNameFound := len(matches) > 0
		if userNameFound {
			results[i] = fmt.Sprintf("[USR] %s %s", matches[1], lines[i])
		} else {
			results[i] = lines[i]
		}
	}

	return results
}
